Source/ui_unix/GamePad/GamePadDeviceListener_OSX.cpp

```diff
@@ -0,0 +1,351 @@
+#include "GamePadDeviceListener_OSX.h"
+#include "GamePadUtils.h"
+
+CGamePadDeviceListener::CGamePadDeviceListener(OnInputEvent OnInputEventCallBack, bool filter)
+    : OnInputEventCallBack(OnInputEventCallBack)
+    , m_running(true)
+    , m_filter(filter)
+{
+	m_thread = std::thread([this]() { InputDeviceListenerThread(); });
+}
+
+CGamePadDeviceListener::CGamePadDeviceListener(bool filter)
+    : CGamePadDeviceListener(nullptr, filter)
+{
+}
+
+CGamePadDeviceListener::~CGamePadDeviceListener()
+{
+	m_running = false;
+	CFRunLoopStop(CFRunLoopGetCurrent());
+	IOHIDManagerClose(m_hidManager, 0);
+	if(m_thread.joinable())
+		m_thread.join();
+}
+
+void CGamePadDeviceListener::UpdateOnInputEventCallback(CGamePadDeviceListener::OnInputEvent OnInputEventFunction)
+{
+	OnInputEventCallBack = OnInputEventFunction;
+}
+
+void CGamePadDeviceListener::DisconnectInputEventCallback()
+{
+	OnInputEventCallBack = nullptr;
+}
+
+void CGamePadDeviceListener::SetFilter(bool filter)
+{
+	m_filter = filter;
+}
+
+void CGamePadDeviceListener::UpdateDeviceList()
+{
+}
+
+CFMutableDictionaryRef CGamePadDeviceListener::CreateDeviceMatchingDictionary(uint32 usagePage, uint32 usage)
+{
+	CFMutableDictionaryRef result = CFDictionaryCreateMutable(kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+	if(result == nullptr) throw std::runtime_error("CFDictionaryCreateMutable failed.");
+	if(usagePage != 0)
+	{
+		// Add key for device type to refine the matching dictionary.
+		CFNumberRef pageCFNumberRef = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &usagePage);
+		if(pageCFNumberRef == nullptr) throw std::runtime_error("CFNumberCreate failed.");
+		CFDictionarySetValue(result, CFSTR(kIOHIDDeviceUsagePageKey), pageCFNumberRef);
+		CFRelease(pageCFNumberRef);
+		// note: the usage is only valid if the usage page is also defined
+		if(usage != 0)
+		{
+			CFNumberRef usageCFNumberRef = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &usage);
+			if(usageCFNumberRef == nullptr) throw std::runtime_error("CFNumberCreate failed.");
+			CFDictionarySetValue(result, CFSTR(kIOHIDDeviceUsageKey), usageCFNumberRef);
+			CFRelease(usageCFNumberRef);
+		}
+	}
+	return result;
+}
+
+void CGamePadDeviceListener::InputValueCallbackStub(void* context, IOReturn result, void* sender, IOHIDValueRef valueRef)
+{
+	auto device_info = reinterpret_cast<DeviceInfo*>(context);
+	if(!device_info->OnInputEventCallBack)
```

Comment:
This tests the std::function's pointer validity and not the actual std::function's value (which can be set to nullptr by DisconnectInputEventCallback). This should be `if(!(*device_info->OnInputEventCallBack))`.